encodings/protobuf/README.md

```diff
@@ -0,0 +1,178 @@
+has significant utility.  This directory contains that canonical
+encoding.
+
+## Semantic Type Considerations
+
+### Primitive Types
+
+The primitive types are mapped as follows
+
+| OCSF Type   | Protobuf Type           |
+| ----------- | ----------------------- |
+| `boolean_t` | `bool`                  |
+| `float_t`   | `float`                 |
+| `integer_t` | `int32`                 |
+| `long_t`    | `int64`                 |
+| `string_t`  | `string`                |
+| `json_t`    | `google.protobuf.Value` |
+
+The mapping is straightforward, except to note that `json_t` and the
+bare (un-extended) `object` type are represented by well-known
+protobuf types that model JSON, specifically
+[Value](https://protobuf.dev/reference/protobuf/google.protobuf/#value)
+and
+[Struct](https://protobuf.dev/reference/protobuf/google.protobuf/#struct)
+respectively.
+
+### Optionality
+
+Most attributes in OCSF are not required; for non-array primitive
+types, the corresponding protobuf descriptor will mark them explicity
+as `optional`.  This has implications for most SDKs generated by
+`protoc`.  For example, in Go, the field of the struct becomes of
+pointer type, adding an extra indirection with the benefit that the
+presence or absence of the field can be consistently determined.
+
+Required attributes in OCSF are not marked as optional.  In most
+cases, an SDK will still permit distinguishing between present and
+absent values because the zero value is not a valid value anyway.  For
+example, `class_id` is required and non-zero; if the application finds
+it to have a zero value, the distinction between missing and invalid
+is not significant because missing _is_ invalid for required types.
+
+### Other Types
+
+Non-primitive types are mapped to the appropriate underlying
+primitive type, with the following exceptions:
+
+#### Date Time
+
+The `datetime_t` type is mapped to the well-known
+[Timestamp](https://protobuf.dev/reference/protobuf/google.protobuf/#timestamp)
+protobuf object.
+
+#### Enumerated Values
+
+Enumerated **integer** values are built using protobuf's `enum`
+declaration.  Because OCSF often (not always) refines the set of
+enumerands on a per-event basis, the protobuf encoding supports that
+by making the enumeration types sub-names of the message types.
+Furthermore, because Protobuf uses "C++" scoping rules, in which
+enumerated values are in the same scope as the enum type itself
+(rather than children of the type), we mangle the enumerand names as
+well.
+
+For example:
+```
+message FileActivity {
+    enum ActionId {
+        UNKNOWN_ActionId = 0;  // The action was unknown. The <code>disposition_id</code>
+                               // attribute may still be set to a non-unknown value, for
+                               // example 'Count', 'Uncorrected', 'Isolated',
+                               // 'Quarantined' or 'Exonerated'.
+        ALLOWED_ActionId = 1;  // The activity was allowed. The
```

Comment:
I put the generated proto through the [Buf](https://buf.build/docs/lint/overview) linting tool and two rules trigger with your enum naming.
Here are two example printouts:
```
ocsf.proto:26:9:Enum value name "ALLOWED_ActionId" should be prefixed with "ACTION_ID_".
ocsf.proto:26:9:Enum value name "ALLOWED_ActionId" should be UPPER_SNAKE_CASE, such as "ALLOWED_ACTION_ID".
```

The fix would be instead of naming enum values like `UNKNOWN_ActionId`, it should be `ACTION_ID_ALLOWED`.
You don't necessarily have to fix this if there are good reasons for doing it this way.

Otherwise, there were a couple other rule triggers so you might want to run Buf yourself.

Reply:
:til: Thanks, I was not aware of a protobuf linting tool!